pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,340 @@
+		},
+		getLabelNames(),
+	)
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(allocatableGaugeVec)
+	crmetrics.Registry.MustRegister(requestsGaugeVec)
+	crmetrics.Registry.MustRegister(limitsGaugeVec)
+	crmetrics.Registry.MustRegister(overheadGaugeVec)
+}
+
+func getLabelNames() []string {
+	return []string{
+		resourceType,
+		nodeName,
+		nodeProvisioner,
+		nodeZone,
+		nodeArchitecture,
+		nodeCapacityType,
+		nodeInstanceType,
+		nodePhase,
+		nodeLabels,
+	}
+}
+
+type Controller struct {
+	KubeClient client.Client
+	LabelsMap  map[types.NamespacedName][]prometheus.Labels
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		KubeClient: kubeClient,
+		LabelsMap:  make(map[types.NamespacedName][]prometheus.Labels),
+	}
+}
+
+// Reconcile executes a termination control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("nodemetrics").With("node", req.Name))
+	// Retrieve node from reconcile request
+	node := &v1.Node{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to node deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				c.deleteGauges(labels)
+			} else {
+				logging.FromContext(ctx).Debugf("Failed to delete gauge: failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after node labels are updated
+	if labelSlice, ok := c.LabelsMap[req.NamespacedName]; ok {
+		c.deleteGauges(labelSlice)
+	}
+	c.LabelsMap[req.NamespacedName] = []prometheus.Labels{}
+
+	if err := c.updateGauges(ctx, node); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to update gauges: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	err := controllerruntime.
+		NewControllerManagedBy(m).
+		Named("nodemetrics").
+		For(&v1.Node{}).
+		Watches(
+			// Reconcile all nodes related to a provisioner when it changes.
+			&source.Kind{Type: &v1alpha5.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				nodes := &v1.NodeList{}
+				if err := c.KubeClient.List(ctx, nodes, client.MatchingLabels(map[string]string{v1alpha5.ProvisionerNameLabelKey: o.GetName()})); err != nil {
+					logging.FromContext(ctx).Debugf("Failed to list nodes when mapping expiration watch events, %s", err.Error())
+					return requests
+				}
+				for _, node := range nodes.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
+				}
+				return requests
+			}),
+		).
+		Watches(
+			// Reconcile node when a pod assigned to it changes.
```

Comment:
How about the following? I want to disambiguate how this works a little more in the codebase.
```suggestion
			// Reconcile all nodes where pods have changed
```

Reply:
In the current design the controller reconcile one node at a time. Do we need to reconcile all nodes when there are pods that have changed?

Reply:
IIUC, since the condition below does `name != ""` the code will functionally queue up requests to all nodes that have pod changes where the nodeName is not defined. This is intended and may be functionally the same, but is a slightly different process from enqueueing a request to one node when any pod on that node changes. 
```
				if name := o.(*v1.Pod).Spec.NodeName; name != "" {
```
You could also make it
`// Reconcile nodes where pods have changed`